File: Programacao-III/aula3/Quadrado.java
package aula3;

public class Quadrado extends Rectangulo {
    Ponto bottomLeft, topRight;
    double side;
    
    public Quadrado() {
        this.center = new Ponto();
        this.bottomLeft = new Ponto();
        this.topRight = new Ponto();
        this.side = 0;
    }

    public Quadrado(Rectangulo r) {
        this.center = r.center;
        this.bottomLeft = r.getbottomLeft();
        this.topRight = r.getTopRight();
        this.side = Math.round(r.area());
    }

    /**
     * 
     * @param bottomLeft
     * @param topRight
     */
    public Quadrado(Ponto bottomLeft, Ponto topRight) {
        Ponto temp = new Ponto(bottomLeft.getX(), topRight.getY());

        assert bottomLeft.distanceTo(temp) == topRight.distanceTo(temp);

        this.center = bottomLeft.halfWayTo(topRight);
        this.bottomLeft = bottomLeft;
        this.topRight = topRight;
        this.side = bottomLeft.distanceTo(temp);
    }

    /**
     * 
     * @param side
     */
    public Quadrado(double side){
        this.center = new Ponto();
        this.side = side;
    }
    
    /**
     * 
     * @param p1
     * @param p2
     * @param side
     */
    public Quadrado(double p1, double p2, double side){
        this.center = new Ponto(p1,p2);
        this.side = side;
        this.bottomLeft = new Ponto(p1-side/2, p2-side/2);
        this.topRight = new Ponto(p1+side/2, p2+side/2);
    }

    /**
     * 
     * @param q2
     */
    public Quadrado(Quadrado q2){
        this.center = q2.center;
        this.side = q2.side;
    }

    @Override
    public double area() {
        return Math.pow(side, 2);
    }

    @Override
    public double perimeter() {
        return 4 * side;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Quadrado) {
            Quadrado temp = (Quadrado)obj;
            return this.center.equals(temp.center) && side == temp.side;
        } else {
            return super.equals(obj);
        }
    }

    @Override
    public String toString() {
        return String.format("Quadrado = { center : %s, bottomLeft : %s, topRight : %s, side : %.2f, area : %.2f, perimeter : %.2f }",
            center, bottomLeft, topRight, side, area(), perimeter()
        );
    }
}

File: 2ºSemestre/LFA/Projeto/geometrics-lfa-07-master/pygame/Quadrado.py
import abc
from Forma import *
import math
from Ponto import *
import pygame

class Quadrado(Forma):
    def __init__(self, *args):

        #if there are no args
        if len(args) == 0 :
            self.center = Ponto(0,0)
            self.bottomLeft = Ponto(0,0)
            self.topRight = Ponto(0,0)
            self.side = 0
            self.color = (0,0,0)
            self.left = 0
            self.top = 0

        #if the args are two dots
        elif(isinstance(args[0],Ponto) and isinstance(args[1],Ponto) and isinstance(args[2],tuple)):
            temp = Ponto(args[0].get_x(), args[1].get_y())

            assert args[0].__distanceTo__(temp) == args[1].__distanceTo__(temp)

            self.center =  args[0].__halfWayTo__(args[1])
            self.bottomLeft = args[0]
            self.topRight = args[1]
            self.side = args[0].__distanceTo__(temp)
            self.color = args[2]
            self.top = self.bottomLeft.get_x()
            self.left = self.bottomLeft.get_y() + self.side

        #in case that the only args is the side
        elif(isinstance(args[0],int)):
            self.center = Ponto(0,0)
            self.side = args[0]
            self.color = args[1]
            self.top = 0 - self.side/2
            self.left = 0 + self.side/2


        #in case that are provided three int values
        elif(isinstance(args[0],int) and isinstance(args[1],int) and isinstance(args[2],int) and isinstance(args[3],tuple)):
            self.center = Ponto(args[0],args[1])
            self.side = args[2]
            self.bottomLeft = Ponto(args[0]-args[2]/2,args[1]-args[2]/2)    #Ponto(0,-250) BottomLeft
            self.topRight = Ponto(args[0]+args[2]/2, args[1] + args[2]/2)   #Ponto(750,250)
            self.color = args[3]
            self.top = self.topRight.get_y() - self.side              
            self.left = self.bottomLeft.get_x()




        

    def __setCenter__(self,center):
        self.center = center

    def __getCenter__(self):
        return self.center

    def __getbottomLeft__(self):
        return self.bottomLeft

    def __gettopRight__(self):
        return self.topRight
    
    def __getColor__(self):
        return self.color
    
    def area(self):
        return math.pow(self.side,2)

    def perimeter(self):
        return 4 * self.side

    def __intercept__(self, other):
        if(self.bottomLeft.get_x() >= other.topRight.get_y() or self.topRight.get_x() <= other.bottomLeft.get_x() or self.topRight.get_y() <= other.bottomLeft.get_y() or self.bottomLeft.get_y() >= other.topRight.get_y()):
            return False
        else:
            return True

    def __draw__(self, surface):
        toDrawRect = pygame.Rect(self.left, self.top, self.side, self.side)
        pygame.draw.rect(surface, self.color, toDrawRect)



